packages/language/src/parser/parser-state.ts

```diff
@@ -301,6 +345,66 @@ export class ParserState {
     }
     return this.consume(element, kind, tokenType);
   }
+
+  private mapMatch(map: RuleMap, tokenType: TokenType): number | undefined {
+    return [
+      tokenType.tokenTypeIdx ?? -1,
+      ...(tokenType.CATEGORIES ?? []).map((cat) => cat.tokenTypeIdx ?? -1),
+    ].find((index) => map.has(index));
```

Comment:
Suggestion: Transforming this into multiple arrrays seems quite slow to me. We can do this better with a guard and a loop:


```suggestion
    const tokenTypeIdx = tokenType.tokenTypeIdx ?? -1;
    if (map.has(tokenTypeIdx) {
      return tokenTypeIdx;
    }
    if (tokenType.CATEGORIES) {
      for (const category of tokenType.CATEGORIES) {
        const categoryIdx = category.tokenTypeIdx ?? -1;
        if (map.has(categoryIdx) {
          return categoryIdx;
        }
      }
    }
    return undefined;
```